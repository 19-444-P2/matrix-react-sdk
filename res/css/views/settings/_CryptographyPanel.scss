.mx_CryptographyPanel {
    display: table;
    padding-left: 0;
}

.mx_CryptographyPanel > li {
    display: table-row;
}

.mx_CryptographyPanel > li > label,
.mx_CryptographyPanel > li > span {
    display: table-cell;
    padding-right: 1em;
}

.mx_CryptographyPanel_importExportButtons .mx_AccessibleButton {
    margin-right: 10px;
}

.mx_CryptographyPanel_importExportButtons {
    margin-bottom: 15px;
}
