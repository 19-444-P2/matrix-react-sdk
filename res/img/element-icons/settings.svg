<svg width="20" height="20" viewBox="0 0 20 20" fill="none" xmlns="http://www.w3.org/2000/svg">
<path fill-rule="evenodd" clip-rule="evenodd" d="M17.3625 7.2875C17.5625 7.8125 18.075 8.1625 18.6375 8.1625C19.3875 8.1625 20 8.775 20 9.525V10.475C20 11.225 19.3875 11.8375 18.6375 11.8375C18.075 11.8375 17.5625 12.1875 17.3625 12.7125C17.346 12.7538 17.3294 12.7958 17.3128 12.838C17.2538 12.9876 17.1932 13.1413 17.125 13.2875C16.8875 13.8 17 14.4 17.4 14.8C17.9375 15.325 17.9375 16.1875 17.4 16.725L16.725 17.4C16.2 17.9375 15.3375 17.9375 14.8 17.4C14.4125 17 13.8 16.8875 13.2875 17.125C13.1 17.2125 12.9125 17.2875 12.7125 17.3625C12.1875 17.5625 11.8375 18.075 11.8375 18.6375C11.8375 19.3875 11.225 20 10.475 20H9.525C8.775 20 8.1625 19.3875 8.1625 18.6375C8.1625 18.075 7.8125 17.5625 7.2875 17.3625C7.24617 17.346 7.20423 17.3294 7.16195 17.3128C7.01243 17.2538 6.85867 17.1932 6.7125 17.125C6.2 16.8875 5.6 17 5.2 17.4C4.675 17.9375 3.8125 17.9375 3.275 17.4L2.6 16.725C2.0625 16.2 2.0625 15.3375 2.6 14.8C3 14.4125 3.1125 13.8 2.875 13.2875C2.7875 13.1 2.7125 12.9125 2.6375 12.7125C2.4375 12.1875 1.925 11.8375 1.3625 11.8375C0.6125 11.8375 0 11.225 0 10.475V9.525C0 8.775 0.6125 8.1625 1.3625 8.1625C1.925 8.1625 2.4375 7.8125 2.6375 7.2875C2.67694 7.16129 2.72634 7.04005 2.77627 6.91751C2.80546 6.84587 2.83483 6.77379 2.8625 6.7C3.1 6.1875 2.9875 5.5875 2.5875 5.1875C2.05 4.6625 2.05 3.8 2.5875 3.2625L3.275 2.6C3.8 2.0625 4.6625 2.0625 5.2 2.6C5.5875 3 6.2 3.1125 6.7125 2.875C6.9 2.7875 7.0875 2.7 7.2875 2.6375C7.8125 2.4375 8.1625 1.925 8.1625 1.3625C8.1625 0.6125 8.775 0 9.525 0H10.475C11.225 0 11.8375 0.6125 11.8375 1.3625C11.8375 1.9375 12.1875 2.4375 12.7125 2.6375C12.7538 2.65403 12.7958 2.67056 12.838 2.68723C12.9876 2.74617 13.1413 2.80679 13.2875 2.875C13.8 3.1125 14.4 3 14.8 2.6C15.325 2.0625 16.1875 2.0625 16.725 2.6L17.4 3.275C17.9375 3.8 17.9375 4.6625 17.4 5.2C17 5.5875 16.8875 6.2 17.125 6.7125C17.2125 6.9 17.2875 7.0875 17.3625 7.2875ZM10 15C7.2375 15 5 12.7625 5 10C5 7.2375 7.2375 5 10 5C12.7625 5 15 7.2375 15 10C15 12.7625 12.7625 15 10 15Z" fill="black"/>
</svg>
