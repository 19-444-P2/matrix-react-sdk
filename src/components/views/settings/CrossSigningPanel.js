/*
Copyright 2019 The Matrix.org Foundation C.I.C.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

import React from 'react';

import MatrixClientPeg from '../../../MatrixClientPeg';
import { _t } from '../../../languageHandler';
import sdk from '../../../index';
import Modal from '../../../Modal';

export default class CrossSigningPanel extends React.PureComponent {
    constructor(props) {
        super(props);
        this.state = this._getUpdatedStatus();
    }

    _getUpdatedStatus() {
        // XXX: Add public accessors if we keep this around in production
        const cli = MatrixClientPeg.get();
        const crossSigning = cli._crypto._crossSigningInfo;
        const secretStorage = cli._crypto._secretStorage;
        const crossSigningPublicKeysOnDevice = crossSigning.getId();
        const crossSigningPrivateKeysInStorage = crossSigning.isStoredInSecretStorage(secretStorage);
        const secretStorageKeyInAccount = secretStorage.hasKey();

        return {
            crossSigningPublicKeysOnDevice,
            crossSigningPrivateKeysInStorage,
            secretStorageKeyInAccount,
        };
    }

    _bootstrapSecureSecretStorage = async () => {
        try {
            const InteractiveAuthDialog = sdk.getComponent("dialogs.InteractiveAuthDialog");
            await MatrixClientPeg.get().bootstrapSecretStorage({
                authUploadDeviceSigningKeys: async (makeRequest) => {
                    const { finished } = Modal.createTrackedDialog(
                        'Cross-signing keys dialog', '', InteractiveAuthDialog,
                        {
                            title: _t("Send cross-signing keys to homeserver"),
                            matrixClient: MatrixClientPeg.get(),
                            makeRequest,
                        },
                    );
                    const [confirmed] = await finished;
                    if (!confirmed) {
                        throw new Error("Cross-signing key upload auth canceled");
                    }
                },
            });
            this.setState(this._getUpdatedStatus());
        } catch (e) {
            console.error(e);
        }
    }

    render() {
        const AccessibleButton = sdk.getComponent("elements.AccessibleButton");
        const {
            crossSigningPublicKeysOnDevice,
            crossSigningPrivateKeysInStorage,
            secretStorageKeyInAccount,
        } = this.state;

        return (
            <div>
                <table className="mx_CrossSigningPanel_statusList"><tbody>
                    <tr>
                        <td>{_t("Cross-signing public keys:")}</td>
                        <td>{crossSigningPublicKeysOnDevice ? _t("on device") : _t("not found")}</td>
                    </tr>
                    <tr>
                        <td>{_t("Cross-signing private keys:")}</td>
                        <td>{crossSigningPrivateKeysInStorage ? _t("in secret storage") : _t("not found")}</td>
                    </tr>
                    <tr>
                        <td>{_t("Secret storage public key:")}</td>
                        <td>{secretStorageKeyInAccount ? _t("in account data") : _t("not found")}</td>
                    </tr>
                </tbody></table>
                <div className="mx_CrossSigningPanel_buttonRow">
                    <AccessibleButton kind="primary" onClick={this._bootstrapSecureSecretStorage}>
                        {_t("Bootstrap Secure Secret Storage")}
                    </AccessibleButton>
                </div>
            </div>
        );
    }
}
