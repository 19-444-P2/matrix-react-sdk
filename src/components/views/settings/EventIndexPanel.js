/*
Copyright 2020 The Matrix.org Foundation C.I.C.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

import React from 'react';

import { _t } from '../../../languageHandler';
import Modal from '../../../Modal';
import SettingsStore, {SettingLevel} from "../../../settings/SettingsStore";
import AccessibleButton from "../elements/AccessibleButton";
import {formatBytes} from "../../../utils/FormattingUtils";
import EventIndexPeg from "../../../indexing/EventIndexPeg";

export default class EventIndexPanel extends React.Component {
    constructor() {
        super();

        this.state = {
            eventIndexSize: 0,
            roomCount: 0,
            eventIndexingEnabled:
                SettingsStore.getValueAt(SettingLevel.DEVICE, 'enableEventIndexing'),
        };
    }

    async updateCurrentRoom(room) {
        const eventIndex = EventIndexPeg.get();
        const stats = await eventIndex.getStats();

        this.setState({
            eventIndexSize: stats.size,
            roomCount: stats.roomCount,
        });
    }

    componentWillUnmount(): void {
        const eventIndex = EventIndexPeg.get();

        if (eventIndex !== null) {
            eventIndex.removeListener("changedCheckpoint", this.updateCurrentRoom.bind(this));
        }
    }

    async componentWillMount(): void {
        this.updateState();
    }

    async updateState() {
        let eventIndexSize = 0;
        let roomCount = 0;

        const eventIndex = EventIndexPeg.get();
        const eventIndexingEnabled = SettingsStore.getValueAt(SettingLevel.DEVICE, 'enableEventIndexing');

        if (eventIndex !== null) {
            eventIndex.on("changedCheckpoint", this.updateCurrentRoom.bind(this));

            const stats = await eventIndex.getStats();
            eventIndexSize = stats.size;
            roomCount = stats.roomCount;
        }

        this.setState({
            eventIndexSize,
            roomCount,
            eventIndexingEnabled,
        });
    }

    _onManage = async () => {
        Modal.createTrackedDialogAsync('Message search', 'Message search',
            import('../../../async-components/views/dialogs/eventindex/ManageEventIndex'),
            {
                onFinished: () => {},
            }, null, /* priority = */ false, /* static = */ true,
        );
    }

    _onEnable = async () => {
        await EventIndexPeg.initEventIndex();
        await EventIndexPeg.get().addInitialCheckpoints();
        await EventIndexPeg.get().startCrawler();
        await SettingsStore.setValue('enableEventIndexing', null, SettingLevel.DEVICE, true);
        await this.updateState();
    }

    render() {
        let eventIndexingSettings = null;

        if (EventIndexPeg.get() !== null) {
            eventIndexingSettings = (
                <div>
                    <div className='mx_SettingsTab_subsectionText'>
                        {_t( "Securely cache encrypted messages locally for them " +
                             "to appear in search results, using ")
                        } {formatBytes(this.state.eventIndexSize, 0)}
                        {_t( " to store messages from ")} {this.state.roomCount} {_t("rooms.")}
                    </div>
                    <div>
                        <AccessibleButton kind="primary" onClick={this._onManage}>
                            {_t("Manage")}
                        </AccessibleButton>
                    </div>
                </div>
            );
        } else if (!this.state.eventIndexingEnabled && EventIndexPeg.supportIsInstalled()) {
            eventIndexingSettings = (
                <div>
                    <div className='mx_SettingsTab_subsectionText'>
                        {_t( "Securely cache encrypted messages locally for them to appear in search results.")}
                    </div>
                    <div>
                        <AccessibleButton kind="primary" onClick={this._onEnable}>
                            {_t("Enable")}
                        </AccessibleButton>
                    </div>
                </div>
            );
        } else {
            eventIndexingSettings = (
                <div>
                    {
                        _t( "Riot can't securely cache encrypted messages locally " +
                            "while running in a web browser. Use Riot Desktop for " +
                            "encrypted messages to appear in search results.",
                        )
                    }
                </div>
            );
        }

        return eventIndexingSettings;
    }
}
