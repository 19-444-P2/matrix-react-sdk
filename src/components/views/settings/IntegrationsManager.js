/*
Copyright 2015, 2016 OpenMarket Ltd
Copyright 2019 The Matrix.org Foundation C.I.C.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

import React from 'react';
import PropTypes from 'prop-types';
import sdk from '../../../index';
import MatrixClientPeg from '../../../MatrixClientPeg';
import { _t } from '../../../languageHandler';
import Modal from '../../../Modal';
import dis from '../../../dispatcher';

export default class IntegrationsManager extends React.Component {
    static propTypes = {
        // the source of the integration manager being embedded
        src: PropTypes.string.isRequired,

        // callback when the manager is dismissed
        onFinished: PropTypes.func.isRequired,
    };

    componentDidMount() {
        this.dispatcherRef = dis.register(this.onAction);
        document.addEventListener("keydown", this.onKeyDown);
    }

    componentWillUnmount() {
        dis.unregister(this.dispatcherRef);
        document.removeEventListener("keydown", this.onKeyDown);
    }

    onKeyDown = (ev) => {
        if (ev.keyCode === 27) { // escape
            ev.stopPropagation();
            ev.preventDefault();
            this.props.onFinished();
        }
    };

    onAction = (payload) => {
        if (payload.action === 'close_scalar') {
            this.props.onFinished();
        }
    };

    render() {
        return <iframe src={ this.props.src }></iframe>;
    }
}