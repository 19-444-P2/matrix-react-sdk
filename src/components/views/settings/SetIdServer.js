/*
Copyright 2019 The Matrix.org Foundation C.I.C.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

import request from 'browser-request';
import url from 'url';
import React from 'react';
import {_t} from "../../../languageHandler";
import sdk from '../../../index';
import MatrixClientPeg from "../../../MatrixClientPeg";
import SdkConfig from "../../../SdkConfig";
import Field from "../elements/Field";

/**
 * If a url has no path component, etc. abbreviate it to just the hostname
 *
 * @param {string} u The url to be abbreviated
 * @returns {string} The abbreviated url
 */
function abbreviateUrl(u) {
    if (!u) return '';

    const parsedUrl = url.parse(u);
    // if it's something we can't parse as a url then just return it
    if (!parsedUrl) return u;

    if (parsedUrl.path == '/') {
        // we ignore query / hash parts: these aren't relevant for IS server URLs
        return parsedUrl.host;
    }

    return u;
}

function unabbreviateUrl(u) {
    if (!u) return '';

    let longUrl = u;
    if (!u.startsWith('https://')) longUrl = 'https://' + u;
    const parsed = url.parse(longUrl);
    if (parsed.hostname === null) return u;

    return longUrl;
}

/**
 * Check an IS URL is valid, including liveness check
 *
 * @param {string} isUrl The url to check
 * @returns {string} null if url passes all checks, otherwise i18ned error string
 */
async function checkIdentityServerUrl(url) {
    const parsedUrl = url.parse(isUrl);

    if (parsedUrl.protocol !== 'https:') return _t("Identity Server URL must be HTTPS");

    // XXX: duplicated logic from js-sdk but it's quite tied up in the validation logic in the
    // js-sdk so probably as easy to duplicate it than to separate it out so we can reuse it
    return new Promise((resolve) => {
        request(
            // also XXX: we don't really know whether to hit /v1 or /v2 for this: we
            // probably want a /versions endpoint like the C/S API.
            { method: "GET", url: isUrl + '/_matrix/identity/api/v1' },
            (err, response, body) => {
                if (err) {
                    resolve(_t("Could not connect to Identity Server"));
                } else if (response.status < 200 || response.status >= 300) {
                    resolve(_t("Not a valid Identity Server (status code %(code)s)", {code: response.status}));
                } else {
                    resolve(null);
                }
            },
        );
    });
}

export default class SetIdServer extends React.Component {
    constructor() {
        super();

        let defaultIdServer = abbreviateUrl(MatrixClientPeg.get().getIdentityServerUrl());
        if (!defaultIdServer) {
            defaultIdServer = abbreviateUrl(SdkConfig.get()['validated_server_config']['idServer']) || '';
        }

        this.state = {
            currentClientIdServer: MatrixClientPeg.get().getIdentityServerUrl(),
            idServer: defaultIdServer,
            error: null,
            busy: false,
        };
    }

    _onIdentityServerChanged = (ev) => {
        const u = ev.target.value;

        this.setState({idServer: u});
    };

    _getTooltip = () => {
        if (this.state.busy) {
            const InlineSpinner = sdk.getComponent('views.elements.InlineSpinner');
            return <div>
                <InlineSpinner />
                { _t("Checking server") }
            </div>;
        } else if (this.state.error) {
            return this.state.error;
        } else {
            return null;
        }
    };

    _idServerChangeEnabled = () => {
        return !!this.state.idServer && !this.state.busy;
    };

    _saveIdServer = async () => {
        this.setState({busy: true});

        const fullUrl = unabbreviateUrl(this.state.idServer);

        const errStr = await checkIsUrl(fullUrl);

        let newFormValue = this.state.idServer;
        if (!errStr) {
            MatrixClientPeg.get().setIdentityServerUrl(fullUrl);
            localStorage.removeItem("mx_is_access_token");
            localStorage.setItem("mx_is_url", fullUrl);
            newFormValue = '';
        }
        this.setState({
            busy: false,
            error: errStr,
            currentClientIdServer: MatrixClientPeg.get().getIdentityServerUrl(),
            idServer: newFormValue,
        });
    };

    render() {
        const idServerUrl = this.state.currentClientIdServer;
        let sectionTitle;
        let bodyText;
        if (idServerUrl) {
            sectionTitle = _t("Identity Server (%(server)s)", { server: abbreviateUrl(idServerUrl) });
            bodyText = _t(
                "You are currently using <server></server> to discover and be discoverable by " +
                "existing contacts you know. You can change your identity server below.",
                {},
                { server: sub => <b>{abbreviateUrl(idServerUrl)}</b> },
            );
        } else {
            sectionTitle = _t("Identity Server");
            bodyText = _t(
                "You are not currently using an identity server. " +
                "To discover and be discoverable by existing contacts you know, " +
                "add one below.",
            );
        }

        return (
            <form className="mx_SettingsTab_section mx_SetIdServer" onSubmit={this._saveIdServer}>
                <span className="mx_SettingsTab_subheading">
                    {sectionTitle}
                </span>
                <span className="mx_SettingsTab_subsectionText">
                    {bodyText}
                </span>
                <Field label={_t("Identity Server")}
                    id="mx_SetIdServer_idServer"
                    type="text" value={this.state.idServer} autoComplete="off"
                    onChange={this._onIdentityServerChanged}
                    tooltip={this._getTooltip()}
                />
                <input className="mx_Dialog_primary"
                    type="submit" value={_t("Change")}
                    disabled={!this._idServerChangeEnabled()}
                />
            </form>
        );
    }
}
