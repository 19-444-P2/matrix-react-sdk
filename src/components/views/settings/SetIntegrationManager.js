/*
Copyright 2019 The Matrix.org Foundation C.I.C.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

import React from 'react';
import {_t} from "../../../languageHandler";
import sdk from '../../../index';
import Field from "../elements/Field";
import {IntegrationManagers} from "../../../integrations/IntegrationManagers";

export default class SetIntegrationManager extends React.Component {
    constructor() {
        super();

        const currentManager = IntegrationManagers.sharedInstance().getPrimaryManager();

        this.state = {
            currentManager,
            url: "", // user-entered text
            error: null,
            busy: false,
        };
    }

    _onUrlChanged = (ev) => {
        const u = ev.target.value;
        this.setState({url: u});
    };

    _getTooltip = () => {
        if (this.state.busy) {
            const InlineSpinner = sdk.getComponent('views.elements.InlineSpinner');
            return <div>
                <InlineSpinner />
                { _t("Checking server") }
            </div>;
        } else if (this.state.error) {
            return this.state.error;
        } else {
            return null;
        }
    };

    _canChange = () => {
        return !!this.state.url && !this.state.busy;
    };

    _setManager = async (ev) => {
        // Don't reload the page when the user hits enter in the form.
        ev.preventDefault();
        ev.stopPropagation();

        this.setState({busy: true});

        const manager = await IntegrationManagers.sharedInstance().tryDiscoverManager(this.state.url);
        if (!manager) {
            this.setState({
                busy: false,
                error: _t("Integration manager offline or not accessible."),
            });
            return;
        }

        try {
            await IntegrationManagers.sharedInstance().overwriteManagerOnAccount(manager);
            this.setState({
                busy: false,
                error: null,
                currentManager: IntegrationManagers.sharedInstance().getPrimaryManager(),
                url: "", // clear input
            });
        } catch (e) {
            console.error(e);
            this.setState({
                busy: false,
                error: _t("Failed to update integration manager"),
            });
        }
    };

    render() {
        const AccessibleButton = sdk.getComponent('views.elements.AccessibleButton');

        const currentManager = this.state.currentManager;
        let managerName;
        let bodyText;
        if (currentManager) {
            managerName = `(${currentManager.name})`;
            bodyText = _t(
                "You are currently using <b>%(serverName)s</b> to manage your bots, widgets, " +
                "and sticker packs.",
                {serverName: currentManager.name},
                { b: sub => <b>{sub}</b> },
            );
        } else {
            bodyText = _t(
                "Add which integration manager you want to manage your bots, widgets, " +
                "and sticker packs.",
            );
        }

        return (
            <form className="mx_SettingsTab_section mx_SetIntegrationManager" onSubmit={this._setManager}>
                <div className="mx_SettingsTab_heading">
                    <span>{_t("Integration Manager")}</span>
                    <span className="mx_SettingsTab_subheading">{managerName}</span>
                </div>
                <span className="mx_SettingsTab_subsectionText">
                    {bodyText}
                </span>
                <Field label={_t("Enter a new integration manager")}
                    id="mx_SetIntegrationManager_newUrl"
                    type="text" value={this.state.url} autoComplete="off"
                    onChange={this._onUrlChanged}
                    tooltipContent={this._getTooltip()}
                />
                <AccessibleButton
                    kind="primary_sm"
                    type="submit"
                    disabled={!this._canChange()}
                    onClick={this._setManager}
                >{_t("Change")}</AccessibleButton>
            </form>
        );
    }
}
