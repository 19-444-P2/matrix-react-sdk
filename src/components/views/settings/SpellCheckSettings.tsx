/*
Copyright 2019 New Vector Ltd

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

import React from 'react';
import LanguageDropdown from "../../../components/views/elements/LanguageDropdown";
import AccessibleButton from "../../../components/views/elements/AccessibleButton";
import {_t} from "../../../languageHandler";

interface ExistingSpellCheckLanguageIProps {
    language: string,
    onRemoved(language: string),
};

interface SpellCheckLanguagesIProps {
    languages: Array<string>,
    onLanguagesChange(languages: Array<string>),
};

interface SpellCheckLanguagesIState {
    newLanguage: string,
}

export class ExistingSpellCheckLanguage extends React.Component<ExistingSpellCheckLanguageIProps> {
    _onRemove = (e) => {
        e.stopPropagation();
        e.preventDefault();

        return this.props.onRemoved(this.props.language);
    };

    render() {
        return (
            <div className="mx_ExistingSpellCheckerLanguage">
                <span className="mx_ExistingSpellCheckerLanguage_language">{this.props.language}</span>
                <AccessibleButton onClick={this._onRemove} kind="danger_sm">
                    {_t("Remove")}
                </AccessibleButton>
            </div>
        );
    }
}

export default class SpellCheckLanguages extends React.Component<SpellCheckLanguagesIProps, SpellCheckLanguagesIState> {
    constructor(props) {
        super(props);
        this.state = {
            newLanguage: "",
        }
    }

    _onRemoved = (language) => {
        const languages = this.props.languages.filter((e) => e !== language);
        this.props.onLanguagesChange(languages);
    };

    _onAddClick = (e) => {
        e.stopPropagation();
        e.preventDefault();

        const language = this.state.newLanguage;
        
        if (!language) return;
        if (this.props.languages.includes(language)) return;

        this.props.languages.push(language)
        this.props.onLanguagesChange(this.props.languages);
    };

    _onNewLanguageChange = (language: string) => {
        if (this.state.newLanguage === language) return;
        this.setState({newLanguage: language});
    };

    render() {
        const existingSpellCheckLanguages = this.props.languages.map((e) => {
            return <ExistingSpellCheckLanguage language={e} onRemoved={this._onRemoved} key={e} />;
        });

        let addButton = (
            <AccessibleButton onClick={this._onAddClick} kind="primary">
                {_t("Add")}
            </AccessibleButton>
        );

        return (
            <div className="mx_SpellCheckerLanguages">
                {existingSpellCheckLanguages}
                <form onSubmit={this._onAddClick} noValidate={true}
                      className="mx_mx_SpellCheckerLanguages_new">
                    <LanguageDropdown className="mx_GeneralUserSettingsTab_spellCheckLanguageInput"
                                      value={this.state.newLanguage}
                                      onOptionChange={this._onNewLanguageChange} />
                    {addButton}
                </form>
            </div>
        );
    };
}
