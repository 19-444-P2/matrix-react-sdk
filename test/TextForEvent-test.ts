import './skinned-sdk';

import { textForEvent } from "../src/TextForEvent";
import { MatrixEvent } from "matrix-js-sdk";
import SettingsStore from "../src/settings/SettingsStore";
import { SettingLevel } from "../src/settings/SettingLevel";
import renderer from 'react-test-renderer';

function mockPinnedEvent(
    pinnedMessageIds?: string[],
    prevPinnedMessageIds?: string[],
): MatrixEvent {
    return new MatrixEvent({
        type: "m.room.pinned_events",
        state_key: "",
        sender: "@foo:example.com",
        content: {
            pinned: pinnedMessageIds,
        },
        prev_content: {
            pinned: prevPinnedMessageIds,
        },
    });
}

describe("TextForPinnedEvent", () => {
    SettingsStore.setValue("feature_pinning", null, SettingLevel.DEVICE, true);

    it("should mention sender", () => {
        const event = mockPinnedEvent();
        expect(textForEvent(event)).toBe("@foo:example.com changed the pinned messages for the room.");
    });
});

describe("TextForPinnedEvent (JSX)", () => {
    SettingsStore.setValue("feature_pinning", null, SettingLevel.DEVICE, true);

    it("should mention sender", () => {
        const event = mockPinnedEvent();
        const component = renderer.create(textForEvent(event, true));
        expect(component.toJSON()).toMatchSnapshot();
    });
});
