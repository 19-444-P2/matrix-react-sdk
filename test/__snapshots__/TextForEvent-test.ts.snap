// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TextForPinnedEvent (JSX) should mention sender 1`] = `
<span>
  <span>
    @foo:example.com changed the 
    <a
      onClick={[Function]}
    >
       
      pinned messages
       
    </a>
     for the room.
  </span>
</span>
`;
